digi/xbee/xsocket.py

```diff
@@ -0,0 +1,814 @@
+            address = list(self.__data_received_from_dict)[0]
+            data_received = self.__data_received_from_dict[address][0:bufsize].copy()
+            # Update the number of bytes left for 'address' in the dictionary.
+            self.__data_received_from_dict[address] = self.__data_received_from_dict[address][bufsize:]
+            # If the number of bytes left for 'address' is 0, remove it from the dictionary.
+            if len(self.__data_received_from_dict[address]) == 0:
+                self.__data_received_from_dict.pop(address)
+            self.__data_received_from_dict_lock.release()
+        # Return the data received for 'address'.
+        return data_received, address
+
+    def send(self, data):
+        """
+        Sends data to the socket and returns the number of bytes sent. The socket must be connected to a remote socket.
+        Applications are responsible for checking that all data has been sent; if only some of the data was
+        transmitted, the application needs to attempt delivery of the remaining data.
+
+        Args:
+            data (Bytearray): the data to send.
+
+        Returns:
+            Integer: the number of bytes sent.
+
+        Raises:
+            ValueError: if the data to send is ``None``.
+            ValueError: if the number of bytes to send is ``0``.
+            XBeeException: if the connection with the XBee device is not open.
+            XBeeSocketException: if the socket is not valid.
+            XBeeSocketException: if the socket is not open.
+        """
+        self.__send(data, False)
+
+    def sendall(self, data):
+        """
+        Sends data to the socket. The socket must be connected to a remote socket. Unlike ``send()``, this method
+        continues to send data from bytes until either all data has been sent or an error occurs. None is returned
+        on success. On error, an exception is raised, and there is no way to determine how much data, if any, was
+        successfully sent.
+
+        Args:
+            data (Bytearray): the data to send.
+
+        Raises:
+            TimeoutException: if the send status response is not received in the configured timeout.
+            ValueError: if the data to send is ``None``.
+            ValueError: if the number of bytes to send is ``0``.
+            XBeeException: if the connection with the XBee device is not open.
+            XBeeSocketException: if the socket is not valid.
+            XBeeSocketException: if the send status is not ``SUCCESS``.
+            XBeeSocketException: if the socket is not open.
+        """
+        self.__send(data)
+
+    def sendto(self, data, address):
+        """
+        Sends data to the socket. The socket should not be connected to a remote socket, since the destination socket
+        is specified by ``address``.
+
+        Args:
+            data (Bytearray): the data to send.
+            address (Tuple): the address (host name and port) of the s
```

Comment:
The end of the sentence is missing.

`address (Tuple): the address (host name and port) of the destination socket.`